frame/clearing-house/src/lib.rs

```diff
@@ -957,27 +1068,130 @@ pub mod pallet {
+						unrealized_pnl: pnl,
+						unrealized_funding: <Self as Instruments>::unrealized_funding(
+							&market, &position,
+						)?,
+					};
+
+					summary
+						.margin_requirement_maintenance
+						.try_add_mut(&info.margin_requirement_maintenance)?;
+					summary.base_asset_value.try_add_mut(&info.base_asset_value)?;
+					summary.unrealized_pnl.try_add_mut(&info.unrealized_pnl)?;
+					summary.unrealized_funding.try_add_mut(&info.unrealized_funding)?;
+					summary.positions_summary.push((market, position, info));
+				}
+			}
+
+			Ok(summary)
+		}
+
+		/// Fully liquidates the user's positions until his account is brought back above the MMR
+		///
+		/// ## Storage modifications
+		///
+		/// - Updates the [`markets`](Markets) of closed positions (according to changes in
+		///   [`Self::close_position_in_market`]).
+		/// - Removes closed [`positions`](Positions)
+		/// - Updates the user's account [`collateral`](Collateral)
+		///
+		/// ## Returns
+		///
+		/// The total liquidation fee
+		fn fully_liquidate_account(
+			user_id: &T::AccountId,
+			summary: AccountSummary<T>,
+		) -> Result<T::Balance, DispatchError> {
+			let mut collateral = Self::get_collateral(user_id).unwrap_or_else(Zero::zero);
+			let mut margin: T::Decimal = Self::updated_balance(
+				&collateral,
+				&summary.unrealized_pnl.try_add(&summary.unrealized_funding)?,
+			)?
+			.into_decimal()?;
+			let AccountSummary::<T> {
+				margin_requirement_maintenance: mut margin_requirement,
+				base_asset_value,
+				mut positions_summary,
+				..
+			} = summary;
+
+			ensure!(margin < margin_requirement, Error::<T>::SufficientCollateral);
+
+			let mut positions = BoundedVec::<Position<T>, T::MaxPositions>::default();
+			let maximum_fee = Self::full_liquidation_penalty().try_mul(&margin)?;
+			let mut fees = T::Balance::zero();
+			// Sort positions from greatest to lowest margin requirement
+			positions_summary.sort_by_key(|(_, _, info)| info.margin_requirement_maintenance.neg());
+			for (mut market, position, info) in positions_summary {
+				if margin < margin_requirement {
+					Self::close_position_in_market(
+						&position,
+						info.direction,
+						&mut market,
+						info.base_asset_value.into_balance()?,
+					)?;
+					Markets::<T>::insert(&position.market_id, market);
+
+					let base_asset_value_share =
+						info.base_asset_value.try_div(&base_asset_value)?;
+					let fee_decimal = base_asset_value_share.try_mul(&maximum_fee)?;
+					margin.try_sub_mut(&fee_decimal)?;
+					margin_requirement.try_sub_mut(&info.margin_requirement_maintenance)?;
+					fees.try_add_mut(&fee_decimal.into_balance()?)?;
+					collateral = Self::updated_balance(
+						&collateral,
+						&info
+							.unrealized_pnl
+							.try_add(&info.unrealized_funding)?
+							.try_sub(&fee_decimal)?,
+					)?;
+				} else {
+					positions
+						.try_push(position)
+						.expect("resulting vector is at most as large as the original; qed");
```

Comment:
Wouldn't be good to create a test to ensure this property always hold?

I know that it is the expected behavior, I'm just commenting to make sure that in the future this property continues to be valid regardless of changes in the code base.

(We must prevent panics at all costs)

Reply:
Yeah, that's fair. Tbh, the `positions_summary` field in `AccountSummary` isn't constrained to be a `BoundedVec` with the same parameters, so it could be the case in the future that a larger vector than the number of maximum positions gets used as input. I'll just keep the error checking then